### DISCUSSION TITLE
Equivalent to "tone()" function

### DISCUSSION THREAD

**Discussion by JasterV** [Rahix/avr-hal#426] [Q&A] [answered]
Hello there :)

I'm new to arduino and I wanted to start learning it by using Rust since I have extensive experience with the language, but not really in electronics. That is a new world to me.

I'm learning how to use a "passive buzzer" and how to adjust the tone by changing the frequency of the output Pin.
In theory, in order for it to work, I should use a PWM pin, set the duty at 50% and adjust the frequency to the Hz I want so I can play musical notes.

I'm honestly having a hard time understanding how to do this using `avr_hal`. 
The PWM Pin type only implements `set_duty`.

I see that in `embedded_hal` there is a [Pwm trait](https://docs.rs/embedded-hal/latest/embedded_hal/trait.Pwm.html) with a `set_period` which is supposed to adjust the frequency!! Is there any way I can use the `embedded_hal` library along with `avr_hal`? Will it work?

Thank you!

**Comment by JasterV** [comment]
Coming to the basic formula `1 / frequency = Period`
I assume by using this formula that if I want to produce 500hz I will need to set a period of 0.002 seconds (or 2 milliseconds)

But I'm afraid its not that simple since there are "timers" involved and stuff like "Prescalers" (About which I don't have much idea how they work, only the basics but its quite confusing)

Would the following code produce 500hz~ ??

```
loop {
    pin.set_high();
    delay_ms(1u8); // 50% duty on 2ms cycle
    pin.set_low();
    delay_ms(1u8);
}
```

Is there a more sofisticated way I can go with by using PWM pins and tweaking the timers? The way the timers work and all the prescale stuff is super confusing to me, if someone wants to share some resources or explain to me as if I was 5 years old I would appreciate it <3

**Comment by Vollbrecht** [comment, marked as answer]
The pwm-trait in embedded-hal 0.2.x s a bit under-specified. That's why it was discarded in the current alpha versions and is under rework.

Understanding how Timer / PWM hardware peripherals work is essential if you want to work with embedded-mcu. Try to study around it but here is a short rundown.

1. It is hardware logic that runs in parallel to your processor.
2. Its just a counter that counts a register up / or low depending on some configurations, in its simplest form. 
3. It can trigger interupts / events with an internal comperator ( e.g if it reaches count X you set in register A it will fire)
4. The timer itself is derived from some higher clock ( depending on processor - main oscillator etc), usually running at a fraction of that speed. ( example Processor runs at 1MHz -> Timer prescaler configured to devide by 8 -> Timer runs at 125kHz.) 
5. You get the PWM for free if the GPIO is connected to the TIMER and the timer allows to directly output its comperator register to an pin. ( e.g example it counts from 0-255 -> you have two triggers -> 127 and 0 ->  on 0 pin goes high , on 127 pin goes low -> a duty cycle of 50%) 

This is just the rough idea with details missing. The best place to learn is directly interacting with the registers itself, witch you can do here. First you should know how the peripheral is build. To do that the best way is to read the datasheet of the processor itself. If you are using an arduino i assume something like an arduino uno that has an Atmega328p processor. Read this https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-7810-Automotive-Microcontrollers-ATmega328P_Datasheet.pdf chapter 14-17 are the different Timer/Pwm peripherals on that board

**Reply by JasterV** [reply]
Thank you so much this was super useful!! Now I understand much better the relation between the timer HZ and the main processor clock HZ and what role the prescaler plays in there, now it makes sense 

I actually have an Elegoo Mega which runs an ATMega 2560 :D
